### PR TITLE
Terrazzo Keyboard

### PR DESCRIPTION
**Pull request by ademey**
## Description

Hi, this PR is for Terrazzo, a pro-micro 40% keyboard that integrates a IS31FL3731 single color LED module. There are staggered and ortho versions of the keyboard. 

For images and a video, please see this [Imgur page](https://imgur.com/a/L8YjlqU).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/terrazzo/config.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
+#ifndef CONFIG_H
+#define CONFIG_H
```
Comment (ridingqwerty) [inline keyboards/terrazzo/config.h:2]:
```suggestion
#pragma once
```

Prefer `#pragma once` here over the old-style include guards.

File: keyboards/terrazzo/config.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include "config_common.h"
+
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x4d4d // MM
+#define PRODUCT_ID      0x545a // TZ
+#define DEVICE_VER      0x0002
+#define MANUFACTURER    MsMustard
+#define PRODUCT         Terrazzo
+#define DESCRIPTION     40% Keyboard with LED Matrix
```
Comment (ridingqwerty) [inline keyboards/terrazzo/config.h:13]:
```suggestion
```

File: keyboards/terrazzo/config.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#define LED_DRIVER_ADDR_1 0x74 
+#define LED_DRIVER_COUNT 1
+#define LED_DRIVER_LED_COUNT 105
+#define LED_MATRIX_ROWS 15
+#define LED_MATRIX_COLS 7
+#define LED_MATRIX_MAXIMUM_BRIGHTNESS 20
+#define LED_DISABLE_WHEN_USB_SUSPENDED true
+
+#endif
+#endif
```
Comment (ridingqwerty) [inline keyboards/terrazzo/config.h:82]:
```suggestion
```

See above re: `#pragma once`, which obviates the need for a closing `#endif`.

File: keyboards/terrazzo/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,69 @@
+// Staggered layout with split spacebars
+// make terrazzo:default
+#include "terrazzo.h"
```
Comment (ridingqwerty) [inline keyboards/terrazzo/keymaps/default/keymap.c:3]:
```suggestion
```

This is redundant as `#include QMK_KEYBOARD_H` will include `"terrazzo.h"`.

File: keyboards/terrazzo/keymaps/ortho/keymap.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// Ortho layout with 2x 2u spacebars
+// Default ortho layout.
+// make terrazzo:othro
+#include "terrazzo.h"
```
Comment (ridingqwerty) [inline keyboards/terrazzo/keymaps/ortho/keymap.c:4]:
```suggestion
```

This is redundant as `#include QMK_KEYBOARD_H` will include `"terrazzo.h"`.

File: keyboards/terrazzo/keymaps/ortho_all/keymap.c [hunk 6]
```diff
@@ -0,0 +1,68 @@
+/* Ortho layout with all 1u
+   make terrazzo:othro_all */
+#include "terrazzo.h"
```
Comment (ridingqwerty) [inline keyboards/terrazzo/keymaps/ortho_all/keymap.c:3]:
```suggestion
```

This is redundant as `#include QMK_KEYBOARD_H` will include `"terrazzo.h"`.

File: keyboards/terrazzo/keymaps/ortho_mit/keymap.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
+// Ortho layout with 1 2u spacebar
+// make terrazzo:othro_mit
+#include "terrazzo.h"
```
Comment (ridingqwerty) [inline keyboards/terrazzo/keymaps/ortho_mit/keymap.c:3]:
```suggestion
```

This is redundant as `#include QMK_KEYBOARD_H` will include `"terrazzo.h"`.

**Review by ridingqwerty — CHANGES_REQUESTED** [review]
Just some outmoded conventions for which I made some suggestions.  The terrazzo effects bit seems sound although I have no way to test that.  Everything looks good otherwise.

File: keyboards/terrazzo/rules.mk [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
+
```
[the remark this answers is not in the document]
Comment (ademey) [reply keyboards/terrazzo/rules.mk]: I was not able to get my keymaps to build with this change. The one thing it was missing was:
```
# MCU name
MCU = atmega32u4
```
All keymaps build fine now.

File: keyboards/terrazzo/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	  	  _______, KC_CAPS, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_F11,  KC_F12,          _______, 
+	  	  _______, _______, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,         CG_TOGG, 
+	  	  _______,          RESET,   _______,          _______,          _______,          _______, _______
+    )
+};
+
+
+void encoder_update_user(uint8_t index, bool clockwise) {
+    terrazzo_scroll_pixel(clockwise);
+    switch(biton32(layer_state)) {
```
Comment (drashna) [inline keyboards/terrazzo/keymaps/default/keymap.c:58]:
```suggestion
    switch(get_highest_layer(layer_state)) {
```

File: keyboards/terrazzo/keymaps/ortho/keymap.c [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		  _______, _______, KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  _______,
+		  _______, KC_CAPS, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_F11,  KC_F12,  _______, _______, 
+		  _______, _______, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, _______, CG_TOGG, 
+		  _______,          RESET,   _______, _______,      _______,          _______,     _______, _______, _______
+  )
+};
+
+void encoder_update_user(uint8_t index, bool clockwise) {
+    terrazzo_scroll_pixel(clockwise);
+    switch(biton32(layer_state)) {
```
Comment (drashna) [inline keyboards/terrazzo/keymaps/ortho/keymap.c:57]:
```suggestion
    switch(get_highest_layer(layer_state)) {
```

File: keyboards/terrazzo/keymaps/ortho_all/keymap.c [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+	  	  _______, _______, KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  _______,
+	  	  _______, KC_CAPS, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_F11,  KC_F12,  _______, _______, 
+	  	  _______, _______, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, _______, CG_TOGG, 
+	  	  _______,          RESET,   _______, _______, _______, _______, _______, _______, _______, _______, _______
+    )
+};
+
+void encoder_update_user(uint8_t index, bool clockwise) {
+    terrazzo_scroll_pixel(clockwise);
+    switch(biton32(layer_state)) {
```
Comment (drashna) [inline keyboards/terrazzo/keymaps/ortho_all/keymap.c:57]:
```suggestion
    switch(get_highest_layer(layer_state)) {
```

File: keyboards/terrazzo/keymaps/ortho_mit/keymap.c [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+	  	  _______, KC_CAPS, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_F11,  KC_F12,  _______, _______, 
+	  	  _______, _______, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, _______, CG_TOGG, 
+	  	  _______,          RESET,   _______, _______, _______,     _______,      _______, _______, _______, _______
+    )
+};
+
+
+void encoder_update_user(uint8_t index, bool clockwise) {
+    terrazzo_scroll_pixel(clockwise);
+    switch(biton32(layer_state)) {
```
Comment (drashna) [inline keyboards/terrazzo/keymaps/ortho_mit/keymap.c:55]:
```suggestion
    switch(get_highest_layer(layer_state)) {
```

**Review by drashna — CHANGES_REQUESTED** [review]
Very cool! 

However, for all of the `c` and `h` files, could you add a GPL2+ compatible license header?

**Review by drashna — APPROVED** [review]

**Review by ridingqwerty — APPROVED** [review]

**Pull request merged**